### PR TITLE
add CLEVR dataset

### PR DESCRIPTION
**Pull request by pmeier**
Addresses #5108.

cc @pmeier @bjuncek

### CODE REVIEW

**Review by NicolasHug — COMMENTED** [review]
Thanks @pmeier , I made a few comments below but this looks good

File: torchvision/prototype/datasets/utils/_internal.py
Comment (pmeier) [reply torchvision/prototype/datasets/utils/_internal.py:124]: This enables chained `getattr` calls. In turn, this allows me to use this filter function

```python
images_dp = Filter(images_dp, path_comparator("parent.name", config.split))
```

rather than writing a custom function that extracts the name of the parent folder.

Comment (NicolasHug) [reply torchvision/prototype/datasets/utils/_internal.py:124]: I'll leave it up to you to decide, as the entire logic of `path_comparator` and `path_accessor` is honestly too complex for me to handle ATM.

But as rule of thumb, I tend to avoid changes to helper functions when they just only address one single use-case.

File: torchvision/datasets/clevr.py
Comment (pmeier) [reply torchvision/datasets/clevr.py:24]: > In other datasets like DTD, the `download` parameter comes before the transforms parameters. I feel like we should be consistent here.

This is inconsistent across datasets:

https://github.com/pytorch/vision/blob/5c9c8352ece2d5924017bc000f40e391eb06a482/torchvision/datasets/caltech.py#L39-L41

https://github.com/pytorch/vision/blob/5c9c8352ece2d5924017bc000f40e391eb06a482/torchvision/datasets/voc.py#L74-L77

From quickly looking through the source, it seems `download` last is more prevalent. So if we want to standardize here, I would go for that.

> I also wouldn't mind introducing the `*` for keyword-only parameters in these new datasets TBH.

The prototype datasets will no longer have a `download` flag, since the data is managed by `torchvision`. For everything else, we already have them as keyword only there:

https://github.com/pytorch/vision/blob/5c9c8352ece2d5924017bc000f40e391eb06a482/torchvision/prototype/datasets/_api.py#L60-L66

Comment (NicolasHug) [reply torchvision/datasets/clevr.py:24]: > From quickly looking through the source, it seems download last is more prevalent. So if we want to standardize here, I would go for that

Ok, that sounds good. All other PRs that I've reviewed about the FLAVA datasets, `download` comes before though. So we'll need to update them before merging, and modify `Food101` as well as `DTD`.

> The prototype datasets will no longer have a download flag, since the data is managed by torchvision. For everything else, we already have them as keyword only there:

Sorry for being unclear, I was suggesting to use the `*` for these newly-introduced datasets, even if they're not in the `prototype` area.

File: torchvision/prototype/datasets/utils/_internal.py
Comment (pmeier) [reply torchvision/prototype/datasets/utils/_internal.py:124]: > as the entire logic of `path_comparator` and `path_accessor` is honestly too complex for me to handle ATM.

The complexity stems from the fact that we cannot use lambdas or local function together with the datapipes, since they cannot be serialized safely. Quick explanation:

- `path_accessor`: When using datapipes you often have path-handle-tuples for files. `path_accessor` takes such a tuple, turns the the first item into a `pathlib.Path` and accesses it according to the input. So `path_accessor("parent.name")` is equivalent to

    ```python
    def extract_folder_name(data):
        path = pathlib.Path(data[0])
        return path.parent.name
    ```

    This is useful when you want to merge two datapipes and need a function that generates merge keys.
- `path_comparator`: This is one layer on top of `path_accessor` by providing an equality comparison for the extract path information. For example `Filter(scenes_dp, path_comparator("name", f"CLEVR_{config.split}_scenes.json"))` will select all files where the file name matches the second attribute. (This is what I used to refactor your suggestion about the regex).

> But as rule of thumb, I tend to avoid changes to helper functions when they just only address one single use-case.

It is true, that this currently only addresses this, but I already needed to make the same changes to `getitem` a while back. So I'm guessing if I don't do it now, I'll have to do it in the future anyway, but then I also have to remember that I need to go back to here and fix this.

File: torchvision/datasets/clevr.py
Comment (pmeier) [reply torchvision/datasets/clevr.py:24]: xref #5164

**Review by NicolasHug — APPROVED** [review]
Thanks a lot @pmeier , LGTM!

**Pull request merged**
